### ISSUE TITLE
`yolo predict detect save_txt=True` appends results from different images with same name to the same output .txt file

### ISSUE THREAD

**Issue by jim-meyer** [ultralytics/ultralytics#22531]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

My images and labels have a nested directory structure like this:
```
./Dir1/SubDirA/Image1.jpg
./Dir1/SubDirA/Image2.jpg
./Dir1/SubDirA/…
./Dir1/SubDirA/ImageN.jpg

./Dir1/SubDirB/Image1.jpg
./Dir1/SubDirB/Image2.jpg
./Dir1/SubDirB/…
./Dir1/SubDirB/ImageM.jpg

./Dir2/SubDirC/Image1.jpg
./Dir2/SubDirC/Image2.jpg
./Dir2/SubDirC/…
./Dir2/SubDirC/ImageK.jpg
```

There are often more subduers than shown above (e.g. `./Dir2/SubDirC/SubDirD/SubDire/SomeImage.jpg`) but the structure above is simple enough to illustrate the problem I’m having.
The important thing is that each image’s file name may *not* be unique but the path to the image *is* unique (ofc).

When I use `yolo predict detect imgsz=... model=... source=/home/ubuntu/test_images/**/*.* data=... save_txt=True` all of the resulting `.txt` files get put into a single directory. E.g. `ultralytics/runs/detect/MyProjectName/labels`.
*Problem*: I cannot tell which .jpg a given .txt corresponds to.

*Bug*: Worse, the predictions for `./Dir1/SubDirA/Image1.jpg` get saved to Image1.txt. Then the predictions for `./Dir1/SubDirB/Image1.jpg` get *appended to Image1.txt*. So `runs/detect/MyProjectName/labels/Image1.txt` ends up with all of the predictions for all Image1.jpg files!
Silently appending predictions from one image(s) to predictions for another image seems like a bug.

Is there a simpler solution to my problem that I’m overlooking? Certainly other folks must store train/test/validation images in a hierarchical directory structure too. 

Note: It is not feasible for all images to be stored in a single directory. I’m working with datasets with 10s of 1000s of images. Putting that many files in a single directory is a hard “no” both for performance and for manageability. And doing so would break my labeling, training and evaluation pipelines.

### Environment

I'm using `yolo predict detect ...` in the ultralytics/ultralytics:8.3.218-python-export docker container.

### Minimal Reproducible Example

```
EXISTING_IMAGE=<insert filename>.jpg
mkdir -p images/Dir1/SubDirA
mkdir -p images/Dir1/SubDirB
mkdir -p images/Dir1/SubDirC
cp ${EXISTING_IMAGE} images/Dir1/SubDirA
cp ${EXISTING_IMAGE} images/Dir1/SubDirB
cp ${EXISTING_IMAGE} images/Dir1/SubDirC
yolo predict detect model=… source=images/**/*.jpg data=… save_txt=True
cat runs/detect/…/labels/*.txt
```
At this point the output will show the same detection repeated 3 times, one for each of the three images.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jim-meyer, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help get things moving quickly — an Ultralytics engineer will also follow up with you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Thank you for sharing your directory structure and steps — to help us reproduce precisely, could you also include:
-The exact CLI command used (as a single line), including all arguments
-The exact `ultralytics` and `torch` versions from your environment
-The OS and shell you’re using (e.g., bash/zsh) and whether the glob was quoted or unquoted
-A small sample set (2–3 files) that collide by basename and the resulting `runs/detect/.../labels` outputs showing the unexpected append behavior
-The full run directory path printed by the CLI (e.g., `runs/detect/exp123`) and its structure

You can print versions with:
```bash
yolo version
python -c "import ultralytics, torch; print('ultralytics', ultralytics.__version__, 'torch', torch.__version__)"
```

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the clear report — once we have the additional details above and confirmation on the latest version check, we’ll be able to reproduce and investigate this more quickly 🙏🛠️

**Comment by glenn-jocher** [comment]
Thanks for the clear repro—this is a real bug in the CLI: `predictor.write_results()` builds label filenames from `p.stem` only, so identically named files in different folders collide, and `Results.save_txt()` appends to the same path, merging outputs; see the `write_results` reference and `save_txt` reference for current behavior: [engine/predictor.py – write_results](https://docs.ultralytics.com/reference/engine/predictor/), [engine/results.py – save_txt](https://docs.ultralytics.com/reference/engine/results/).

Workarounds now:
1) Use the Python API to preserve your directory tree in `labels/` so filenames are unique.
```python
from pathlib import Path
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
src_root = Path("/home/ubuntu/test_images")  # set to your dataset root
for r in model.predict(source=str(src_root / "**/*.*"), stream=True):
    rel = Path(r.path).resolve().relative_to(src_root)
    out = Path("runs/detect/myproj/labels") / rel.with_suffix(".txt")
    out.parent.mkdir(parents=True, exist_ok=True)
    r.save_txt(out, save_conf=True)
```
2) Pure CLI fallback: run one subdirectory at a time (distinct `project/name`) to avoid collisions until we ship a fix.

Next steps: we’ll keep this open and adjust `write_results()` to either preserve relative paths under `labels/` or use a collision-free scheme. If you can, please paste the output of `yolo checks` from inside your container so we can record exact versions. For `save_txt`/CLI behavior details, the arguments are summarized in the [Predict mode docs](https://docs.ultralytics.com/modes/predict/).

**Comment by jim-meyer** [comment]
Thanks @glenn-jocher . A bug fix is appreciated but I can go with the custom python script for the time being. I really like how simple that script is. That speaks well to well how well factored the python code is.

Output from `yolo checks` as requested:
```
root@d0968de407d2:/ultralytics# yolo checks
Ultralytics 8.3.218 🚀 Python-3.11.10 torch-2.8.0+cu128 CPU (Intel Core i9-9880H 2.30GHz)
Setup complete ✅ (12 CPUs, 19.5 GB RAM, 155.3/243.0 GB disk)

OS                     Linux-6.10.14-linuxkit-x86_64-with-glibc2.36
Environment            Docker
Python                 3.11.10
Install                git
Path                   /ultralytics/ultralytics
RAM                    19.53 GB
Disk                   155.3/243.0 GB
CPU                    Intel Core i9-9880H 2.30GHz
CPU count              12
GPU                    None
GPU count              None
CUDA                   None

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.9.4>=3.3.0
opencv-python-headless ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.2>=1.4.1
torch                  ✅ 2.8.0<=2.8.0,>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,<=2.8.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.1.1
polars                 ✅ 1.34.0
ultralytics-thop       ✅ 2.0.17>=2.0.0
```

**Comment by glenn-jocher** [comment]
Thanks for the checks output—captured; we’ll keep this open and update here once the CLI fix lands to prevent label filename collisions. If you prefer a flat `labels/` dir without mirroring folders, you can make filenames unique by hashing the relative path:

```python
from pathlib import Path
from ultralytics import YOLO
import hashlib

model = YOLO("yolo11n.pt")
root = Path("/home/ubuntu/test_images")
out_dir = Path("runs/detect/myproj/labels_flat")

for r in model.predict(source=str(root / "**/*.*"), stream=True):
    rel = Path(r.path).resolve().relative_to(root)
    h = hashlib.sha1(rel.as_posix().encode()).hexdigest()[:8]
    out = out_dir / f"{rel.stem}__{h}.txt"
    out.parent.mkdir(parents=True, exist_ok=True)
    if out.exists():  # avoid appending across reruns
        out.unlink()
    r.save_txt(out, save_conf=True)
```

If you’re willing to validate the fix on your dataset when a PR is up, that would help.

**Comment by jim-meyer** [comment]
Whoa, what?! Why is a bug like this being closed?

**Comment by glenn-jocher** [comment]
This was auto-closed by our stale triage (the `Stale` label), not because it’s “fixed” or dismissed; if you can still reproduce on the current `ultralytics` version, please open a fresh report via the [bug report template](https://github.com/ultralytics/ultralytics/issues/new?template=bug-report.yml) and link back to this thread so it pops back onto the team’s active queue.
